File: Episode12-Leaderboard/Monogame/Bullet.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using MonoGame.Extended;
using System;

namespace Shmup
{
    internal class Bullet
    {
        /// <summary>
        /// Mob class for generating meteors from top of window
        /// </summary>
        #region Properties
        private RectangleF rectangle;
        private CircleF circle;
        public RectangleF Rectangle  { get { return rectangle; } } 
        public CircleF Circle { get { return circle; } }
        #endregion
        #region Class variables
        private bool active;
        private Texture2D bulletImg;
        private float speedY;
        private float radius;
        #endregion
        #region Constructor
        public Bullet(Texture2D bulletimg, string align = "centre")
        {
            bulletImg = bulletimg;
            /// Create a rectangle based on Player position, assuming centre firing point
            rectangle = new RectangleF
            (
                Player.Rectangle.X + (Player.Rectangle.Width - bulletImg.Width)  / 2,
                Player.Rectangle.Y - bulletImg.Height,
                bulletImg.Width,
                bulletImg.Height
            );
            /// Adjust rectangle for left or right firing point
            if (align == "left")
                rectangle.X = Player.Rectangle.X;
            else if (align == "right")
                rectangle.X = Player.Rectangle.Right - bulletImg.Width;
            radius = MathHelper.Min(Rectangle.Width, Rectangle.Height) / 2;
            circle = new CircleF(Rectangle.Center, radius);
            speedY = 1000;
            if (Shared.Debug) speedY = speedY / 2;
            active = true; // Not Active allows removal of this object in main Update
        }
        #endregion
        #region Class Methods
        public bool Update(float dt)
        {
            rectangle.Y -= speedY * dt;
            circle.Center.Y = Rectangle.Y;
            if (Rectangle.Y < 0 - Rectangle.Height)
                active = false;
            return active;
        }
        public void Draw(SpriteBatch spriteBatch)
        {
            if (Shared.Debug)
            {
                spriteBatch.DrawCircle(circle: Circle, sides: 20, color: Color.Red, thickness: 1);
                spriteBatch.DrawRectangle(Rectangle, Color.Yellow);
            }
            else
                spriteBatch.Draw(bulletImg, Rectangle.Position, Color.White);
        }
        #endregion
    }
}

File: Episode12-Leaderboard/Love2D/bullet.lua
local Class = require("lib.Class")
local Bullet = Class:derive("Bullet")
local Rectangle = require "lib.Rectangle"
local Circle = require "lib.Circle"
local Shared = require "shared"

function Bullet:new(Player, sprite, align)
	--[[ class constructor, takes player and a sprite]]
	self.sprite = sprite
	local posX = Player.rect.x + Player.rect.w / 2 - sprite:getWidth() / 2
	if align == 'left' then
		posX = Player.rect.left
	elseif align == 'right' then
		posX = Player.rect.right - sprite:getWidth()
	end
	self.speedY = 1000
	if Shared.debug then
		self.speedY = self.speedY / 2
	end
	self.radius = math.min( sprite:getWidth() / 2, sprite:getHeight() / 2)
	-- rectangle collider
	self.rect = Rectangle(posX,
						  Player.rect.y - sprite:getHeight() + 10,
						  sprite:getWidth(),
						  sprite:getHeight())
	-- circle collider
	self.circle = Circle(self.rect.x + self.radius,
						 self.rect.y + self.radius,
						 self.radius)
	self.active = true
end

function Bullet:update(dt)
	self.rect:updatePosition(0, self.speedY * dt * -1) -- negative as bullets go up
	self.circle:updatePosition(0, self.speedY * dt * -1)
	if self.rect.top <= 0 then
		self.active = false 
	end
	return self.active -- when false, bullet can be set to nil
end

function Bullet:draw()
	love.graphics.setColor(Shared.WHITE)
	love.graphics.draw(self.sprite, self.rect.x, self.rect.y)
	if Shared.debug then
		love.graphics.setColor(Shared.RED)
		love.graphics.circle('line', self.circle.x, self.circle.y, self.circle.radius)
		love.graphics.setColor(Shared.YELLOW)
		love.graphics.rectangle("line", self.rect.x, self.rect.y, self.rect.w, self.rect.h)
	end
end

return Bullet